src/scenic/syntax/veneer.py

```diff
@@ -111,1000 +132,1341 @@
+    X = toVector(X, '"follow F from X for D" with X not a vector')
+    D = toScalar(D, '"follow F from X for D" with D not a number')
+    pos = F.followFrom(X, D)
+    orientation = F[pos]
+    return OrientedPoint._with(position=pos, parentOrientation=orientation)
+
+def VisibleFromOp(region, base):
+    """The :grammar:`<region> visible from <point>` operator."""
+    region = toType(region, Region, '"X visible from Y" with X not a Region')
+    if not isA(base, Point):
+        raise TypeError('"X visible from Y" with Y not a Point')
+    return region.intersect(base.visibleRegion)
 
 def CanSee(X, Y):
-	"""The :scenic:`{X} can see {Y}` polymorphic operator.
+    """The :scenic:`{X} can see {Y}` polymorphic operator.
+
+    Allowed forms::
+
+        <point> can see <point>
+    """
+    if not isA(X, Point):
+        raise TypeError('"X can see Y" with X not a Point, OrientedPoint, or Scenic Object')
 
-	Allowed forms::
+    assert currentScenario is not None
 
-		<point> can see <object>
-		<point> can see <vector>
-	"""
-	if not isinstance(X, Point):
-		raise RuntimeParseError('"X can see Y" with X not a Point')
-	if isinstance(Y, Point):
-		return X.canSee(Y)
-	else:
-		Y = toVector(Y, '"X can see Y" with Y not a vector')
-		return X.visibleRegion.containsPoint(Y)
+    if currentScenario._sampledObjects is None:
+        raise InvalidScenarioError('"X can see Y" cannot be evaluated before sample time')
 
+    occludingObjects = tuple(obj for obj in currentScenario._sampledObjects if obj.occluding \
+                         and X is not obj and Y is not obj)
+
+    occludingObjects = toDistribution(occludingObjects)
+
+    @distributionFunction
+    def canSeeHelper(X, Y, occludingObjects):
+        for obj in occludingObjects:
+            assert not needsSampling(obj)
+
+        return X.canSee(Y, occludingObjects=occludingObjects)
+
+    return canSeeHelper(X,Y, occludingObjects)
 ### Specifiers
 
 def With(prop, val):
-	"""The :grammar:`with <property> <value>` specifier.
+    """The :grammar:`with <property> <value>` specifier.
 
-	Specifies the given property, with no dependencies.
-	"""
-	return Specifier(prop, val)
+    Specifies the given property, with no dependencies.
+    """
+    return Specifier("With", {prop: 1}, {prop: val})
 
 def At(pos):
-	"""The :grammar:`at <vector>` specifier.
+    """The :grammar:`at <vector>` specifier.
 
-	Specifies :prop:`position`, with no dependencies."""
-	pos = toVector(pos, 'specifier "at X" with X not a vector')
-	return Specifier('position', pos)
+    Specifies :prop:`position`, with no dependencies.
+    """
+    pos = toVector(pos, 'specifier "at X" with X not a vector')
+    return Specifier("At", {'position': 1}, {'position': pos})
 
 def In(region):
-	"""The :grammar:`in/on <region>` specifier.
+    """The :grammar:`in <region>` specifier.
+
+    Specifies :prop:`position` and :prop:`parentOrientation`, with 
```

Comment:
Doesn't specify `parentOrientation` unless the region has a preferred orientation.